tests/phpunit/tests/option/option.php

```diff
@@ -312,4 +312,151 @@ public function data_option_autoloading() {
 			array( 'autoload_false', false, 'no' ),
 		);
 	}
+
+	/**
+	 * @ticket 22192
+	 *
+	 * @covers ::add_option
+	 */
+	public function test_add_option_with_value_of_false_should_store_false_in_the_cache() {
+		add_option( 'foo', false );
+		$a = wp_cache_get( 'alloptions', 'options' );
+		$this->assertSame( false, $a['foo'] );
+	}
+
+	/**
+	 * @ticket 22192
+	 *
+	 * @covers ::add_option
+	 */
+	public function test_add_option_with_value_of_false_should_store_empty_string_in_the_database() {
+		add_option( 'foo', false );
+
+		// Delete cache to ensure we pull from the database.
+		wp_cache_delete( 'alloptions', 'options' );
+
+		$this->assertSame( '', get_option( 'foo' ) );
+	}
+
+	/**
+	 * @ticket 22192
+	 *
+	 * @covers ::add_option
+	 * @covers ::update_option
+	 *
+	 * @dataProvider data_update_option_type_juggling
+	 *
+	 * @param mixed $old_value One of the values to compare.
+	 * @param mixed $new_value The other value to compare.
+	 */
+	public function test_update_option_should_hit_cache_when_loosely_equal_to_existing_value_and_cached_values_are_faithful_to_original_type( $old_value, $new_value ) {
```

Comment:
I think this test as written is trying to confirm behavior which I think is wrong. From what I can tell, this is essentially ensuring that if an option is set and then updated to a value that is loosely equal, but not strictly equal (e.g., `true` and `1`), then the option will not be updated. However, I think that's wrong because there are valid reasons why you may want to update an option value from one type to another.

Instead, what I think this ticket is trying to ensure, is that if you try to update an option with the same value that is strictly equal, then a database query will not be run because `update_option` will find that the value already exists in cache, whereas now, the cached value is not strictly equal for some reason.

Maybe I'm misunderstanding?

Reply:
I run `test_update_option_should_hit_cache_when_loosely_equal_to_existing_value_and_cached_values_are_faithful_to_original_type` unit test for old and new/updated `update_option` function with dataset. Here is the observation.

For all this dataset value it hit DB query in old `update_option` and it will not run any additional query in updated function.

```
('1', 1)
('1', 1.0)
('1', true)
(1, 1)
(1, 1.0)
(1, true)
(1.0, 1)
(1.0, 1.0)
(1.0, true)
(true, 1)
(true, 1.0)
(true, true)
('0', 0)
('0', 0.0)
('0', false)
(0, 0)
(0, 0.0)
(0, false)
(0.0, 0)
(0.0, 0.0)
(0.0, false)
```

For all this dataset values it update the option in old `update_option` function but not in updated function.
```
('0', false)
(0, false)
(0.0, false)
```

Reply:
> Instead, what I think this ticket is trying to ensure, is that if you try to update an option with the same value that is strictly equal, then a database query will not be run because update_option will find that the value already exists in cache, whereas now, the cached value is not strictly equal for some reason.

That's right yeah, though the ticket also says:
> [The functions] should be smart enough to recognize that `"1" == 1 == true` and `"0" == 0 == false`, and that any numeric string is also equal to a properly cast integer.

So that's where the `if loosely equal, don't update` comes from.

> From what I can tell, this is essentially ensuring that if an option is set and then updated to a value that is loosely equal, but not strictly equal (e.g., true and 1), then the option will not be updated. However, I think that's wrong because there are valid reasons why you may want to update an option value from one type to another.

This also occurred to me when looking at the ticket and PR, and I had the same opinion as you but I didn't have any example use cases in mind.

Reply:
Spending some more time with this issue and it does seem like I misunderstood the details of what the current behavior is and what we are trying to achieve here. I've created [an example PR](https://github.com/WordPress/wordpress-develop/pull/5250) that includes the basic unit tests from this PR along with a complete set of truthy and falsey use cases. You can see that there are [30 use cases](https://github.com/WordPress/wordpress-develop/actions/runs/6241644487/job/16944188040?pr=5250) where the tests fail (i.e, updates are happening when we don't expect them)—all of which are currently falsey use cases. So it seems to me that the original issue as reported might already be mostly fixed.

Also, (except for some falsey use cases) the current behavior already shows that you are not able to update the value of an option to a loosely equivalent scalar value of a different type (e.g., from `'1'` to `true`), So the goal here seems to be to further reduce the times where falsey values are being compared but result in a DB update when values are loosely equivalent.

If there are falsey use cases that MUST keep the current behavior, then we should also add a unit test to ensure those are not effected by this, or future changes.

Reply:
If it would be helpful, I think it would be fine to add a passing version of these tests into trunk prior to any changes here. That said, I think it would be ok to do it all as a part of this ticket. What is not super clear to me from what I'm observing from the tests in my PR is whether any of the use cases that are currently failing can actually be fixed without introducing backwards compatibility breaks. 

Ideally, we should have a complete set of use cases that all pass with the expected behavior of not doing a DB update when the new value is loosely equal to the existing value. For any cases where we MUST do DB updates even on loosely equivalent values, we should also have tests that confirm that behavior with some inline docs explaining why these can't be changed.

Reply:
@joemcgill I think either way works, but committing these tests standalone to `trunk` would clarify (also to future reviewers of this PR) that those changes _continue_ to pass. Adding them here does not make it clear whether those tests would have also passed prior, i.e. whether they cover new or existing behavior.

Reply:
> committing these tests standalone to trunk would clarify (also to future reviewers of this PR) that those changes continue to pass. Adding them here does not make it clear whether those tests would have also passed prior, i.e. whether they cover new or existing behavior.

+1

Also that by having tests already in `trunk`, if any of the tests are _intended_ to change, those changes will be clear to see in this PR and changeset diff when this PR is committed. That means if there are any BC breaks we didn't catch during review, finding the problem doesn't mean to looking at _every_ dataset, but only the _changed_ ones.

Reply:
Now that a set of tests confirming current behavior are in trunk, via https://core.trac.wordpress.org/changeset/56648, is this test still needed?